### PR TITLE
ref/max_error_creation

### PR DESCRIPTION
**Pull request by alhiwatan**
Refactor creation of `MaxError` object.

### CODE REVIEW

File: applovin_max/ios/Classes/AppLovinMAX.m [hunk 1]
```diff
@@ -916,14 +916,8 @@ - (void)sendErrorEventWithName:(NSString *)name
...
-    NSDictionary *body = ( error ) ?
-    @{@"adUnitId": adUnitIdentifier,
-      @"errorCode" : @(error.code),
-      @"errorMessage" : error.message,
-      @"waterfall": [self createAdWaterfallInfo: error.waterfall]}
-    :
-    @{@"adUnitId": adUnitIdentifier,
-      @"errorCode" : @(MAErrorCodeUnspecified)};
+    NSMutableDictionary *body = [@{@"adUnitId" : adUnitIdentifier} mutableCopy];
+    [body addEntriesFromDictionary: [self errorInfo: error]];
```
Comment (thomasmso) [inline applovin_max/ios/Classes/AppLovinMAX.m:920]: I would try to keep parity with React Native:

https://github.com/AppLovin/AppLovin-MAX-React-Native/blob/4ec90a6f9875dcef880bd6d1cead0c52492e199f/ios/AppLovinMAX.m#L1343

**Comment by alhiwatan** [comment]
Add adLoadFailedInfoForAd and adDisplayFailedInfoForAd.  Please review.

File: applovin_max/ios/Classes/AppLovinMAX.m [hunk 2]
```diff
@@ -1601,6 +1578,26 @@ + (CGSize)adViewSizeForAdFormat:(MAAdFormat *)adFormat
              @"waterfall": [self createAdWaterfallInfo: ad.waterfall]};
 }
 
+- (NSDictionary<NSString *, id> *)adLoadFailedInfoForAd:(NSString *)adUnitIdentifier withError:(MAError *)error
```
Comment (thomasmso) [inline applovin_max/ios/Classes/AppLovinMAX.m:1581]: ObjC naming -> `adLoadFailedInfoForAdUnitIdentifier:(NSString *)adUnitIdentifier error:(MAError *)error`

File: applovin_max/ios/Classes/AppLovinMAX.m [hunk 3]
```diff
@@ -1601,6 +1578,26 @@ + (CGSize)adViewSizeForAdFormat:(MAAdFormat *)adFormat
...
+    :
+    @{@"adUnitId": adUnitIdentifier,
+      @"code" : @(MAErrorCodeUnspecified)};
+}
+
+- (NSDictionary<NSString *, id> *)adDisplayFailedInfoForAd:(MAAd *)ad withError:(MAError *)error
+{
+    NSMutableDictionary *body = [NSMutableDictionary dictionaryWithCapacity: 2];
+    body[@"ad"] = [self adInfoForAd: ad];
+    body[@"error"] = [self adLoadFailedInfoForAd: ad.adUnitIdentifier withError: error];
```
Comment (thomasmso) [inline applovin_max/ios/Classes/AppLovinMAX.m:1597]: Does `adInfoForAd:` or `adLoadFailedInfoForAdUnitIdentifier:error:` return `nil`? If not, you can just return the dictionary `@{}` literal like

```
return @{
    @"ad" : [self adInfoForAd: ad],
    @"error" : [self adLoadFailedInfoForAdUnitIdentifier: ad.adUnitIdentifier error: error]
};
```

File: applovin_max/example/android/app/build.gradle [hunk 4]
```diff
@@ -1,3 +1,9 @@
+plugins {
```
Comment (thomasmso) [inline applovin_max/example/android/app/build.gradle:1]: I think you might need to merge master into this branch so your changes from https://github.com/AppLovin/AppLovin-MAX-Flutter/pull/188 do not show in this PR?

File: applovin_max/android/src/main/java/com/applovin/applovin_max/AppLovinMAX.java [hunk 5]
```diff
@@ -957,28 +957,8 @@ else if ( mAppOpenAds.containsKey( adUnitId ) )
...
-            else
-            {
-                params.put( "errorCode", MaxErrorCode.UNSPECIFIED );
-            }
-
-            fireCallback( name, params, channel );
+        try {
+            fireCallback( name, getAdLoadFailedInfo( adUnitId, error ) );
         }
         catch ( Throwable ignored ) { }
```
Comment (thomasmso) [inline applovin_max/android/src/main/java/com/applovin/applovin_max/AppLovinMAX.java:963]: I would remove the `try-catch`. If you insist on keeping it, I'd put it in `fireCallback(...)`, but if there are exceptions being thrown it'd be nice to know what they are and why they happen so I'm leaning towards just removing.

File: applovin_max/android/src/main/java/com/applovin/applovin_max/AppLovinMAX.java [hunk 6]
```diff
@@ -1064,10 +1044,7 @@ else if ( MaxAdFormat.REWARDED == adFormat )
 
         try
         {
-            Map<String, Object> params = getAdInfo( ad );
-            params.put( "errorCode", error.getCode() );
-            params.put( "errorMessage", error.getMessage() );
-            fireCallback( name, params );
+            fireCallback( name, getAdDisplayFailedInfo( ad, error ) );
         }
         catch ( Throwable ignored ) { }
```
Comment (thomasmso) [inline applovin_max/android/src/main/java/com/applovin/applovin_max/AppLovinMAX.java:1049]: Same comment as above.

File: applovin_max/example/android/app/build.gradle (hunk 4 above)
Comment (alhiwatan) [reply applovin_max/example/android/app/build.gradle]: I somehow rebased the master and now had to force push to remove it.

**Comment by alhiwatan** [comment]
Please review this again.

**Review by thomasmso — APPROVED** [review]
Since this is a big change, I would also update changelog to `Refactored error handling logic.` or something like that.

File: applovin_max/ios/Classes/AppLovinMAX.m [hunk 7]
```diff
@@ -1601,6 +1578,26 @@ + (CGSize)adViewSizeForAdFormat:(MAAdFormat *)adFormat
...
+    @{@"adUnitId": adUnitIdentifier,
+      @"code" : @(MAErrorCodeUnspecified)};
+}
+
+- (NSDictionary<NSString *, id> *)adDisplayFailedInfoForAd:(MAAd *)ad withError:(MAError *)error
+{
+    return @{
+        @"ad" : [self adInfoForAd: ad],
+        @"error" : [self adLoadFailedInfoForAdUnitIdentifier: ad.adUnitIdentifier withError: error]
+    };
```
[the remark this answers is not in the document]
Comment (alhiwatan) [reply applovin_max/ios/Classes/AppLovinMAX.m:1598]: Ok, but I like this simple format.

[the remark this answers is not in the document]
Comment (thomasmso) [reply applovin_max/ios/Classes/AppLovinMAX.m:1598]: Same :)

**Pull request merged**